File: tracsmobile/src/main/java/edu/txstate/mobile/tracs/util/Registrar.java
package edu.txstate.mobile.tracs.util;


import android.content.Context;
import android.util.Log;

import com.android.volley.Response;
import com.google.firebase.iid.FirebaseInstanceId;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import edu.txstate.mobile.tracs.AnalyticsApplication;
import edu.txstate.mobile.tracs.BuildConfig;
import edu.txstate.mobile.tracs.R;
import edu.txstate.mobile.tracs.util.http.HttpQueue;
import edu.txstate.mobile.tracs.util.http.listeners.RegisterCallback;
import edu.txstate.mobile.tracs.util.http.requests.DispatchRegistrationRequest;
import edu.txstate.mobile.tracs.util.http.requests.DispatchUnregisterRequest;
import edu.txstate.mobile.tracs.util.http.requests.JwtRequest;

public class Registrar {

    private Map<String, String> registration = new HashMap<>();
    private String jwt;
    private RegisterCallback registerCallback;
    private Response.ErrorListener errorListener;


    private static final String TOKEN_URL = AnalyticsApplication.getContext().getString(R.string.jwt_url);
    private static final String DISPATCH_URL = AnalyticsApplication.getContext().getString(R.string.dispatch_registration);
    private static final String TAG = "Registrar";
    private static Registrar registrar;

    private Registrar() {
        init();
    }

    private void init() {
        Context context = AnalyticsApplication.getContext();
        registration.put("token", FirebaseInstanceId.getInstance().getToken());
        registration.put("platform", "android");
        registration.put("app_id", BuildConfig.APPLICATION_ID);
        registration.put("user_id", AppStorage.get(AppStorage.USERNAME, context));
    }

    public static Registrar getInstance() {
        if (registrar == null) {
            registrar = new Registrar();
        }
        return registrar;
    }

    private JSONObject getJsonRegistration() {
        Registrar.getInstance().init(); //In case the data is out of date or not populated
        return new JSONObject(registration);
    }

    public void registerDevice(RegisterCallback registerCallback, Response.ErrorListener errorListener) {
        this.registerCallback = registerCallback;
        this.errorListener = errorListener;
        HttpQueue requestQueue = HttpQueue.getInstance(AnalyticsApplication.getContext());
        JwtRequest jwtRequest = new JwtRequest(
                TOKEN_URL,
                Registrar.getInstance()::receiveJwt,
                error -> Log.e(TAG, "Could not retrieve JWT from service.")
        );
        requestQueue.addToRequestQueue(jwtRequest, this);
    }

    public void unregisterDevice() {
        HttpQueue requestQueue = HttpQueue.getInstance(AnalyticsApplication.getContext());
        String token = FirebaseInstanceId.getInstance().getToken();
        DispatchUnregisterRequest unregisterRequest = new DispatchUnregisterRequest(DISPATCH_URL, token);
        requestQueue.addToRequestQueue(unregisterRequest, this);
    }

    private void receiveJwt(String jwt) {
        String url = DISPATCH_URL + "?jwt=" + jwt;
        HttpQueue requestQueue = HttpQueue.getInstance(AnalyticsApplication.getContext());
        JSONObject regInfo = Registrar.getInstance().getJsonRegistration();
        DispatchRegistrationRequest registerRequest = new DispatchRegistrationRequest(url, regInfo, registerCallback, errorListener);
        requestQueue.addToRequestQueue(registerRequest, this);
    }
}

File: tracsmobile/src/main/java/edu/txstate/mobile/tracs/util/Registrar.kt
package edu.txstate.mobile.tracs.util

import android.util.Log
import com.android.volley.Response
import com.google.firebase.iid.FirebaseInstanceId
import edu.txstate.mobile.tracs.AnalyticsApplication
import edu.txstate.mobile.tracs.BuildConfig
import edu.txstate.mobile.tracs.R
import edu.txstate.mobile.tracs.util.http.HttpQueue
import edu.txstate.mobile.tracs.util.http.listeners.RegisterCallback
import edu.txstate.mobile.tracs.util.http.requests.DispatchRegistrationRequest
import edu.txstate.mobile.tracs.util.http.requests.DispatchUnregisterRequest
import edu.txstate.mobile.tracs.util.http.requests.JwtRequest
import org.json.JSONObject

class Registrar {
    private val registration = HashMap<String, String>()
    private lateinit var registerCallback: RegisterCallback
    private lateinit var errorListener: Response.ErrorListener

    companion object {
        val TOKEN_URL: String = AnalyticsApplication.getContext().getString(R.string.jwt_url)
        val DISPATCH_URL: String = AnalyticsApplication.getContext().getString(R.string.dispatch_registration)
        val TAG = "Registrar"
        @JvmStatic
        val instance = Registrar()
    }

    init {
        updateRegistration()
    }

    private fun updateRegistration() {
        val context = AnalyticsApplication.getContext()
        registration.put("token", FirebaseInstanceId.getInstance().token ?: "")
        registration.put("platform", "android")
        registration.put("app_id", BuildConfig.APPLICATION_ID)
        registration.put("user_id", AppStorage.get(AppStorage.USERNAME, context))
    }

    private fun getJsonRegistration(): JSONObject {
        updateRegistration()
        return JSONObject(Registrar.instance.registration)
    }

    fun registerDevice(registerCallback: RegisterCallback, errorListener: Response.ErrorListener) {
        this.registerCallback = registerCallback
        this.errorListener = errorListener
        val requestQueue = HttpQueue.getInstance(AnalyticsApplication.getContext())
        val jwtRequest = JwtRequest(
                TOKEN_URL,
                Registrar.instance::receiveJwt,
                { _ -> Log.e(TAG, "Could not retrieve a valid JWT") }
        )
        requestQueue.addToRequestQueue(jwtRequest, this)
    }

    fun unregisterDevice() {
        val requestQueue = HttpQueue.getInstance(AnalyticsApplication.getContext())
        val token = FirebaseInstanceId.getInstance().token
        val unregisterRequest = DispatchUnregisterRequest(DISPATCH_URL, token)
        requestQueue.addToRequestQueue(unregisterRequest, this)
    }

    private fun receiveJwt(jwt: String) {
        val url = "$DISPATCH_URL?jwt=$jwt"
        val requestQueue = HttpQueue.getInstance(AnalyticsApplication.getContext())
        val regInfo = Registrar.instance.getJsonRegistration()
        val registerRequest = DispatchRegistrationRequest(url, regInfo, registerCallback, errorListener)
        requestQueue.addToRequestQueue(registerRequest, this)
    }
}